### PR TITLE
use clap for parsing tests.

### PR DESCRIPTION
**Pull request by astrale-sharp**
add a subtest option (to run only one subtest)

This has proven useful to me while testing, I can remove the subset option or we could make it better (you can specify ranges).

One thing I do not like which should be considered before merging is that the subset selected will be the same for all file selected (it only make sense for when you're testing one file) should I make it an error then?

### CODE REVIEW

**Comment by astrale-sharp** [comment]
bad cherry picking, sorry for that

File: tests/Cargo.toml [hunk 1]
```diff
@@ -16,6 +16,9 @@ tiny-skia = "0.6.2"
 ttf-parser = "0.17"
 unscanny = "0.1"
 walkdir = "2"
+clap = { version = "4.2.1", features = ["derive", "env"] }
+clap_complete = "4.2.0"
```
Comment (laurmaedje) [inline tests/Cargo.toml:20]: `clap_complete` does not seem to be used? also you have an extra newline there :)

File: tests/src/tests.rs [hunk 2]
```diff
@@ -30,10 +29,64 @@ const PDF_DIR: &str = "pdf";
...
+
+#[derive(Debug, Clone, Parser)]
+#[clap(name = "typst-test", author)]
+struct Args {
+    filter: Vec<String>,
+    // runs only the subtest specified
+    #[arg(short, long)]
+    subtest: Option<usize>,
+    #[arg(long, action = SetTrue)]
+    exact: bool,
```
Comment (laurmaedje) [inline tests/src/tests.rs:42]: I think the `action = SetTrue` is not necessary

File: tests/src/tests.rs [hunk 3]
```diff
@@ -30,10 +29,64 @@ const PDF_DIR: &str = "pdf";
...
+
+impl PrintConfig {
+    fn from_arg(arg: &Args) -> Self {
+        Self {
+            syntax: arg.syntax,
+            model: arg.model,
+            frames: arg.frame,
+        }
+    }
+}
```
Comment (laurmaedje) [inline tests/src/tests.rs:71]: I guess we can get rid of this and just use `Args` in its place. Or maybe have the clap stuff on here if there is an equivalent to `serde(flatten)` for clap.

File: tests/src/tests.rs [hunk 4]
```diff
@@ -337,6 +340,7 @@ fn test(
     png_path: &Path,
     ref_path: &Path,
     pdf_path: Option<&Path>,
+    subset_range: Option<usize>,
```
Comment (laurmaedje) [inline tests/src/tests.rs:343]: its not a range, right?

**Comment by astrale-sharp** [comment]
I must have been sleepy and forgot the range stuff !
I'll finish that tonight (most likely) when i get back, don't  worry about merging the other test pr (update) if it's ready, I'll solve the conflicts

Envoyé à partir de Outlook pour Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by astrale-sharp** [comment]
Should be good to go, thanks for the previous review!

**Comment by astrale-sharp** [comment]
Should be good to go! this is not adressed yet :
> One thing I do not like which should be considered before merging is that the subset selected will be the same for all file selected (it only make sense for when you're testing one file) should I make it an error then?

**Comment by laurmaedje** [comment]
> One thing I do not like which should be considered before merging is that the subset selected will be the same for all file selected (it only make sense for when you're testing one file) should I make it an error then?

I think it's fine as is.

**Pull request merged**